### DISCUSSION TITLE
How do I add a new custom `ConfigType`

### DISCUSSION THREAD

**Discussion by dagsir** [dagster-io/dagster#8544] [Q&A]
Hi there! What is the best way to add a new custom `ConfigType`? I want to introduce a "datetime" `ConfigType`, so that strings like '2020-01-01' from config YAML be coerced to Python's datetime objects. I assume I can extend `ConfigType` class itself, but what `ConfigTypeKind` should I use? `ConfigTypeKind.SCALAR` seems appropriate, but then I need to set `self.scalar_kind` attribute...

---

*The question was originally asked in [Dagster Slack](https://dagster.slack.com/archives/C01U954MEER/p1637172399342200?thread_ts=1637172399.342200&cid=C01U954MEER).*

**Comment by yuhan** [comment]
Surfacing [an answer](https://dagster.slack.com/archives/C01U954MEER/p1637222307372200?thread_ts=1637172399.342200&cid=C01U954MEER) from Mikhail Knyazev:
```python
class Datetime(ConfigType):
    """Datetime config type."""

    def __init__(self):
        super().__init__(
            key='datetime',
            kind=ConfigTypeKind.SCALAR,
            given_name='datetime',
        )
        self.scalar_kind = ConfigScalarKind.STRING

    def post_process(self, value: str) -> datetime:
        """Load datetime object from a string.

        :param value: datetime string
        :returns: datetime object
        """
        try:
            datetime_ = dateutil.parser.parse(value)
        except dateutil.parser.ParserError:
            raise PostProcessingError(f'"{value}" is not a valid datetime')
        return datetime_
```

**Comment by C0DK** [comment]
Can we get a more complete example (also is the above up to date?)

**Reply by jamiedemaria** [reply]
@benpankow  are `ConfigTypes` still in use now that we use pydantic configuration?

**Reply by C0DK** [reply]
I am trying to use it with pydantic configuration haha.

The problem is that you need to do something custom to support, for instance, a datetime in a config (which is the issue i had). 

I ended up saving my datetime as a string and then creating logic that validated that it was formatted as expected and creating helpers to extract the actual datetime.. however doing this whenever I want a datetime, seems excessive. I'd prefer to wrap these abstractions into a class (and the ConfigType seemed like the abstraction i needed)

**Reply by jamiedemaria** [reply]
And if you make a ConfigType class like the one in the original post and use that in your config, does it not work? any errors that you get?

**Reply by benpankow** [reply]
Hi @C0DK, right now there's not a way to define a custom config type using the Pydantic-based config models. There's some flexibility in that you could define a custom method on the `Config` class, e.g. something like the following:

```python
from datetime import datetime

import dateutil
from dagster import Config, asset

class MyConfig(Config):
    start_date: str
    end_date: str

    @property
    def start_datetime(self) -> datetime:
        return dateutil.parser.parse(self.start_date)

    @property
    def end_datetime(self) -> datetime:
        return dateutil.parser.parse(self.end_date)

@asset
def my_asset(config: MyConfig):
    print(config.start_datetime)
    print(config.end_datetime)

my_asset(MyConfig(start_date="2024-01-01", end_date="2024-01-05"))
```

That being said this is something I've heard requested with the new config system, so I put up a tracking issue: https://github.com/dagster-io/dagster/issues/19329

**Reply by ydennisy** [reply]
@benpankow Are not all pydantic types supported?

https://docs.pydantic.dev/2.0/usage/types/datetime/

**Comment by C0DK** [comment]
Also can we utilize the [DagsterType](https://docs.dagster.io/concepts/types#associating-dagster-types-with-python-types) for configs?

**Reply by jamiedemaria** [reply]
Dagster types are just for inputs and outputs of ops and assets, not for config
